pkg/driver/node.go

```diff
@@ -0,0 +1,566 @@
+	}
+
+	// A temporary TCP connection is made to the volume to check if its
+	// reachable
+	logrus.Debug("NodeStageVolume: wait for the iscsi target to be ready")
+	if err := ns.waitForVolumeToBeReachable(
+		instance.Spec.ISCSISpec.TargetPortals[0],
+	); err != nil {
+		return nil,
+			status.Error(codes.FailedPrecondition, err.Error())
+	}
+
+	devicePath, err := ns.attachDisk(instance)
+	if err != nil {
+		logrus.Errorf("NodeStageVolume: failed to attachDisk for volume %v, err: %v", req.GetVolumeId(), err)
+		return nil, status.Error(codes.Internal, err.Error())
+	}
+
+	instance.Spec.MountInfo.FSType = fsType
+	instance.Spec.MountInfo.DevicePath = devicePath
+	instance.Spec.MountInfo.Path = stagingPath
+	if err := ns.client.UpdateJivaVolume(instance); err != nil {
+		return nil, status.Error(codes.Internal, err.Error())
+	}
+
+	if err := os.MkdirAll(stagingPath, 0750); err != nil {
+		logrus.Errorf("failed to mkdir %s, error: %v", stagingPath, err)
+		return nil, status.Error(codes.Internal, err.Error())
+	}
+
+	logrus.Info("NodeStageVolume: start format and mount operation")
+	if err := ns.formatAndMount(req, instance.Spec.MountInfo.DevicePath); err != nil {
+		return nil, status.Error(codes.Internal, err.Error())
+	}
+
+	return &csi.NodeStageVolumeResponse{}, nil
+}
+
+// NodeUnstageVolume unmounts the volume from
+// the staging path
+//
+// This implements csi.NodeServer
+func (ns *node) NodeUnstageVolume(
+	ctx context.Context,
+	req *csi.NodeUnstageVolumeRequest,
+) (*csi.NodeUnstageVolumeResponse, error) {
+
+	volID := req.GetVolumeId()
+	if volID == "" {
+		return nil, status.Error(codes.InvalidArgument, "NodeUnstageVolume Volume ID must be provided")
+	}
+
+	if err := ns.client.Set(); err != nil {
+		return nil, status.Error(codes.Internal, err.Error())
+	}
+
+	target := req.GetStagingTargetPath()
+	if len(target) == 0 {
+		return nil, status.Error(codes.InvalidArgument, "Staging target not provided")
+	}
+
+	// Check if target directory is a mount point. GetDeviceNameFromMount
+	// given a mnt point, finds the device from /proc/mounts
+	// returns the device name, reference count, and error code
+	dev, refCount, err := ns.mounter.GetDeviceName(target)
+	if err != nil {
+		msg := fmt.Sprintf("failed to check if volume is mounted: %v", err)
+		return nil, status.Error(codes.Internal, msg)
+	}
+
+	// From the spec: If the volume corresponding to the volume_id
+	// is not staged to the staging_target_path, the Plugin MUST
+	// reply 0 OK.
+	if refCount == 0 {
+		logrus.Infof("NodeUnstageVolume: %s target not mounted", target)
+		return &csi.NodeUnstageVolumeResponse{}, nil
+	}
+
+	if refCount > 1 {
+		logrus.Warningf("NodeUnstageVolume: found %d references to device %s mounted at target path %s", refCount, dev, target)
+	}
+
+	logrus.Debugf("NodeUnstageVolume: unmounting %s", target)
+	err = ns.mounter.Unmount(target)
+	if err != nil {
+		return nil, status.Errorf(codes.Internal, "Could not unmount target %q: %v", target, err)
+	}
+
```

Comment:
If pod restarts after this step, stale iscsi entries will be left out. Since once the path is unmounted, kubernetes will not send the Unstage request again.

Reply:
Please keep a TODO over here to handle it later.